File: frontend/mobile_app/lib/data/http_catalog_repository.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../core/config/app_config.dart';
import '../models/category.dart';
import '../models/product.dart';
import 'backend_dto.dart';
import 'catalog_repository.dart';

/// [CatalogRepository] backed by the Node backend's HTTP API (docs/API.md).
/// Selected when [AppConfig.backendMode] is on; otherwise the app uses
/// [MockCatalogRepository]. The rest of the app is unaware which is in play.
class HttpCatalogRepository implements CatalogRepository {
  HttpCatalogRepository({http.Client? client})
    : _client = client ?? http.Client();

  final http.Client _client;
  static const Duration _timeout = Duration(seconds: 8);

  @override
  Future<List<Category>> categories() async {
    final http.Response res = await _client
        .get(AppConfig.http('/api/categories'))
        .timeout(_timeout);
    if (res.statusCode != 200) return const <Category>[];
    final Map<String, dynamic> body =
        jsonDecode(res.body) as Map<String, dynamic>;
    final List<dynamic> items =
        (body['categories'] as List<dynamic>?) ?? const <dynamic>[];
    return items
        .map((dynamic e) => BackendDto.category(e as Map<String, dynamic>))
        .toList();
  }

  @override
  Future<List<Product>> products({String? categoryId, String? query}) async {
    final Map<String, dynamic> q = <String, dynamic>{
      'limit': 100,
      'category': ?categoryId,
      if (query != null && query.trim().isNotEmpty) 'q': query.trim(),
    };
    final http.Response res = await _client
        .get(AppConfig.http('/api/products', q))
        .timeout(_timeout);
    if (res.statusCode != 200) return const <Product>[];
    final Map<String, dynamic> body =
        jsonDecode(res.body) as Map<String, dynamic>;
    final List<dynamic> items =
        (body['items'] as List<dynamic>?) ?? const <dynamic>[];
    return items
        .map((dynamic e) => BackendDto.fromListItem(e as Map<String, dynamic>))
        .toList();
  }

  @override
  Future<Product?> productById(String id) async {
    final http.Response res = await _client
        .get(AppConfig.http('/api/products/$id'))
        .timeout(_timeout);
    if (res.statusCode != 200) return null;
    return BackendDto.fromDetail(jsonDecode(res.body) as Map<String, dynamic>);
  }

  @override
  Future<List<Product>> recommendations({
    String? gender,
    String? ageRange,
    String? personality,
    String? customerId,
    int limit = 12,
  }) async {
    final Map<String, dynamic> q = <String, dynamic>{
      'limit': limit,
      if (gender != null && gender.isNotEmpty) 'gender': gender,
      if (ageRange != null && ageRange.isNotEmpty) 'ageRange': ageRange,
      if (personality != null && personality.isNotEmpty)
        'personality': personality,
      if (customerId != null && customerId.isNotEmpty) 'customerId': customerId,
    };
    final http.Response res = await _client
        .get(AppConfig.http('/api/recommendations', q))
        .timeout(_timeout);
    if (res.statusCode != 200) return const <Product>[];
    final Map<String, dynamic> body =
        jsonDecode(res.body) as Map<String, dynamic>;
    final List<dynamic> items =
        (body['items'] as List<dynamic>?) ?? const <dynamic>[];
    return items
        .map((dynamic e) => BackendDto.fromListItem(e as Map<String, dynamic>))
        .toList();
  }
}

File: frontend-rn/display_app/src/data/httpCatalogRepository.ts
import { AppConfig } from '../config/appConfig';
import { Category } from '../models/category';
import { Product } from '../models/product';
import { BackendDto } from './backendDto';
import { CatalogRepository } from './catalogRepository';

/// CatalogRepository backed by the Node backend's HTTP API. Selected when
/// AppConfig.backendMode is on. Ported from the Flutter `HttpCatalogRepository`.
export class HttpCatalogRepository implements CatalogRepository {
  private static readonly timeoutMs = 8000;

  private async getJson(url: string): Promise<any | null> {
    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), HttpCatalogRepository.timeoutMs);
    try {
      const res = await fetch(url, { signal: controller.signal });
      if (res.status !== 200) return null;
      return await res.json();
    } catch {
      return null;
    } finally {
      clearTimeout(t);
    }
  }

  async categories(): Promise<Category[]> {
    const body = await this.getJson(AppConfig.http('/api/categories'));
    if (body == null) return [];
    const items = (body.categories as any[]) ?? [];
    return items.map((e) => BackendDto.category(e));
  }

  async products(opts?: { categoryId?: string; query?: string }): Promise<Product[]> {
    const query: Record<string, any> = { limit: 100 };
    if (opts?.categoryId != null) query.category = opts.categoryId;
    if (opts?.query != null && opts.query.trim().length > 0) query.q = opts.query.trim();
    const body = await this.getJson(AppConfig.http('/api/products', query));
    if (body == null) return [];
    const items = (body.items as any[]) ?? [];
    return items.map((e) => BackendDto.fromListItem(e));
  }

  async productById(id: string): Promise<Product | undefined> {
    const body = await this.getJson(AppConfig.http(`/api/products/${id}`));
    if (body == null) return undefined;
    return BackendDto.fromDetail(body);
  }
}
